### PR TITLE
Unidirectional many-to-many relationships

### LINKED ISSUE #3864: Support unidirectional many-to-many relationships through shadow navigations
Support shadow navigation properties such that views over the navigations can function even when a CLR type doesn't have an actual navigation property. This would enable unidirectional many-to-many relationships like this:

```cs
class Post
{
  public int Id { get; set; }
  public ICollection<Tag> Tags { get; set; }
}

class Tag
{
  public int Id { get; set; }
}
```
```cs
builder.Entity<Post>()
  .HasMany(p => p.Tags)
  .WithMany();
```

----

Original issue:

> Are there any plans to support this?
> 
> ```
> table.Include(t => EF.Property<Entity>(t, "Property"))
> ```

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #3864

### CODE REVIEW

**Comment by ajcvickers** [comment]
@dotnet/efteam Creating this as a draft for now since it needs https://github.com/dotnet/efcore/issues/27493 to be implemented before it can be completed. This is needed in order to execute the query for `CollectionEntry.Load`, which uses filtered Include, but can't use the MemberInfo-based lambda overload, since there is no CLR navigation property.

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs [hunk 1]
```diff
@@ -921,11 +921,12 @@ public object GetOrCreateCollection(INavigationBase navigationBase, bool forMate
             ? GetOrCreateCollectionTyped(navigationBase)
             : navigationBase.GetCollectionAccessor()!.GetOrCreate(Entity, forMaterialization);
 
-    private ICollection<object> GetOrCreateCollectionTyped(INavigationBase navigation)
+    private object GetOrCreateCollectionTyped(INavigationBase navigation)
```
Comment (AndriySvyryd) [inline src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs:924]: Update the name

File: src/EFCore/Infrastructure/ModelValidator.cs [hunk 2]
```diff
@@ -115,16 +115,6 @@ public virtual void Validate(IModel model, IDiagnosticsLogger<DbLoggerCategory.M
                         CoreStrings.SkipNavigationNoInverse(
                             skipNavigation.Name, skipNavigation.DeclaringEntityType.DisplayName()));
                 }
-
-                if (skipNavigation.IsShadowProperty())
-                {
-                    throw new InvalidOperationException(
-                        CoreStrings.ShadowManyToManyNavigation(
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ModelValidator.cs:122]: Remove resource

File: src/EFCore/Internal/ManyToManyLoader.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 using Microsoft.EntityFrameworkCore.ChangeTracking.Internal;
 using Microsoft.EntityFrameworkCore.Metadata.Internal;
+using ValueBuffer = Microsoft.EntityFrameworkCore.Storage.ValueBuffer;
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/ManyToManyLoader.cs:6]: Why is this necessary?

File: src/EFCore/Internal/ManyToManyLoader.cs [hunk 4]
```diff
@@ -61,7 +63,8 @@ public virtual async Task LoadAsync(InternalEntityEntry entry, CancellationToken
         // Short-circuit for any null key values for perf and because of #6129
         if (keyValues != null)
         {
-            await Query(entry.Context, keyValues).LoadAsync(cancellationToken).ConfigureAwait(false);
+            await (await Query(entry.Context, keyValues, true, cancellationToken).ConfigureAwait(false))
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/ManyToManyLoader.cs:66]: Split into two lines

File: src/EFCore/Metadata/IClrCollectionAccessor.cs [hunk 5]
```diff
@@ -44,6 +44,30 @@ public interface IClrCollectionAccessor
...
+    /// <summary>
+    ///     Checks whether the value is contained in the collection.
+    /// </summary>
+    /// <param name="collection">The collection.</param>
+    /// <param name="value">The value to check.</param>
+    /// <returns><see langword="true" /> if the value is contained in the collection; <see langword="false" /> otherwise.</returns>
+    bool ContainsStandalone(object collection, object value);
+
+    /// <summary>
+    ///     Removes a value from the collection.
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/IClrCollectionAccessor.cs:64]: Add a note on how these are different from the above

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 6]
```diff
@@ -4118,139 +4118,160 @@ private static bool Contains(IReadOnlyForeignKey? inheritedFk, IReadOnlyForeignK
...
+                    foreach (var (navigation, inverse) in detachedNavigations)
+                    {
+                        navigation.Attach(this, inverseBuilder: inverse);
+                    }
+                }
+            }
+        }
+        else
+        {
+            if (navigationName == null)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs:4256]: Redundant check

File: test/EFCore.SqlServer.FunctionalTests/Query/TPCManyToManyNoTrackingQuerySqlServerTest.cs [hunk 7]
```diff
@@ -15,7 +15,7 @@ public TPCManyToManyNoTrackingQuerySqlServerTest(TPCManyToManyQuerySqlServerFixt
     protected override bool CanExecuteQueryString
         => true;
 
-    [ConditionalFact]
+    [ConditionalFact(Skip = "TODOU: Needs #27493")]
```
Comment (AndriySvyryd) [inline test/EFCore.SqlServer.FunctionalTests/Query/TPCManyToManyNoTrackingQuerySqlServerTest.cs:18]: Remove this and otheres. Also use American spelling

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs:924]: Renamed to `GetOrCreateShadowCollection`

File: src/EFCore/Internal/ManyToManyLoader.cs (hunk 4 above)
Comment (ajcvickers) [reply src/EFCore/Internal/ManyToManyLoader.cs:66]: All of this hack is going away now that the code can use EF.Property for filtered query.

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs [hunk 8]
```diff
@@ -445,44 +445,6 @@ public virtual void Throws_for_conflicting_many_to_one_on_right()
...
-                    nameof(ManyToManyNavPrincipal),
-                    nameof(NavDependent)),
-                Assert.Throws<InvalidOperationException>(
-                    () => modelBuilder.Entity<ManyToManyNavPrincipal>()
-                        .HasMany<NavDependent>( /* leaving empty causes the exception */)
-                        .WithMany(d => d.ManyToManyPrincipals)).Message);
-        }
-
-        [ConditionalFact]
-        public virtual void Throws_for_many_to_many_with_a_shadow_navigation()
```
[the remark this answers is not in the document]
Comment (ajcvickers) [reply test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs:464]: Done.
